### PR TITLE
ref(*): refactor Porter's outputs logic to use the cnab-go library where possible

### PR DESCRIPTION
**Pull request by vdice**
# What does this change
This PR refactors logic in Porter around fetching bundle outputs.  The main theme is the removal of custom Porter code, which was written before support was added in the canonical golang library in https://github.com/deislabs/cnab-go.  Specifically, instead of Porter cache-ing/maintaining an outputs store on the user's local machine in between actions, we read outputs from the updated claims (cnab-go does the work of adding/updating outputs on a claim).

# What issue does it fix
Closes https://github.com/deislabs/porter/issues/634
Closes https://github.com/deislabs/porter/issues/520

# Notes for the reviewer

It is worth calling out that this is a breaking change.  Bundles with dependencies may encounter issues if all dependent bundles aren't updated/rebuilt with the same (future) release of Porter, as the output data structure has changed (drastically simplified).  As this aligns us with the official golang library and enables streamlined support for future outputs functionality and driver support down the line, we've determined it to be a necessary trade-off.  

One subtle change in behavior to consider is this: previously, outputs would be written (locally, as mentioned above) for any/all custom actions (and provided to the user accordingly.)  However, now that we are fetching values from the claim, and outputs are only added/updated on a claim if/when a custom action has `modifies: true` (the non-default), as seen [here](https://github.com/deislabs/cnab-go/blob/master/action/run_custom.go#L57-L61), outputs will only be shown/updated (via `porter instance outputs list` or `porter instance show`) for custom actions with this attribute set to `true`; hence the update to one of the integration test bundles as seen [here](https://github.com/vdice/porter/blob/ref/outputs-cnab-go/examples/exec-outputs/porter.yaml#L10-L16)

# Checklist
- [x] Unit Tests
- [ ] Documentation
  - [x] Documentation Not Impacted

### CODE REVIEW

**Comment by vdice** [comment]
Looking into integration test failure...

**Comment by vdice** [comment]
@carolynvs Now that we were accessing outputs from claims, including for dependencies, I needed to find some way to determine what action was being run when executing a dependency, specifically for the case when the action is `uninstall` (there will no longer be a claim to attempt to access outputs from in this case).  To do so, I added the changes in: https://github.com/deislabs/porter/pull/643/commits/1861a138bed3d14140c495b6c0e3405aace2d46d but I'd love your take on if a different approach would be better.

**Comment by carolynvs** [comment]
Sorry I haven't reviewed it yet! I promise to review it today and give it some thought on how you plumbed in the action name.

File: pkg/porter/run.go [hunk 1]
```diff
@@ -238,32 +242,9 @@ func (p *Porter) ApplyBundleOutputs(opts RunOptions, outputs map[string]string)
...
-						Value:     outputValue,
-					}
-
-					data, err := output.JSONMarshal()
-					if err != nil {
-						return errors.Wrap(err, "unable to marshal output")
-					}
-
-					err = p.FileSystem.WriteFile(outpath, data, 0755)
+					err := p.FileSystem.WriteFile(outpath, []byte(outputValue), 0755)
```
Comment (carolynvs) [inline pkg/porter/run.go]: Now that the output data is stored in the claim, and we aren't storing additional data in the output file in porter home, does it make sense to keep storing the output file in ~/.porter/outputs? I'm wondering what value do we get out of having `ApplyBundleOutputs` or if we could/should switch to having all the commands work off of the outputs in the claim in a follow-up?

File: pkg/porter/outputs.go [hunk 2]
```diff
@@ -110,56 +106,65 @@ func (p *Porter) ListBundleOutputs(opts *OutputListOptions) error {
...
-		value := o.Value
-		// If output is sensitive, substitute local path
-		if o.Sensitive {
-			value = filepath.Join(claimOutputsDir, o.Name)
+	for _, name := range keys {
+		var outputType string
+		valueStr := fmt.Sprintf("%v", outputs[name])
+
+		// Fetch the Output definition from the claim to access metadata, etc.
+		if c.Bundle != nil {
```
[the remark this answers is not in the document]
Comment (vdice) [reply pkg/porter/outputs.go]: Good call, thank you.  Updated.

File: pkg/porter/invoke.go [hunk 3]
```diff
@@ -50,7 +51,7 @@ func (p *Porter) InvokeBundle(opts InvokeOptions) error {
 		return err
 	}
 
-	err = deperator.Execute()
+	err = deperator.Execute(config.ActionCustom)
```
[the remark this answers is not in the document]
Comment (vdice) [reply pkg/porter/invoke.go]: 👍 updated

File: pkg/config/actions.go [hunk 4]
```diff
@@ -6,6 +6,7 @@ const (
 	ActionInstall   Action = "install"
 	ActionUpgrade   Action = "upgrade"
 	ActionUninstall Action = "uninstall"
+	ActionCustom    Action = "custom"
```
[the remark this answers is not in the document]
Comment (vdice) [reply pkg/config/actions.go]: Agreed; updated.

File: pkg/porter/run.go (hunk 1 above)
Comment (vdice) [reply pkg/porter/run.go]: This is writing to the standard path inside the invocation container (`/cnab/app/outputs`).  Indeed, with the changes in this PR, we will no longer be writing/storing outputs on the local user's system (`~/.porter/outputs` will no longer exist)

Comment (carolynvs) [reply pkg/porter/run.go]: Thanks for the clarification! I did indeed have it all mixed up. 😀

**Pull request merged**